python/paddle/v2/fluid/memory_optimization_transpiler.py

```diff
@@ -113,60 +109,130 @@ def _get_diff(self, a, b):
         u = a & b
         return a - u, b - u
 
+    def _has_var(self, block_desc, var_name, is_forward):
+        if is_forward:
+            return block_desc.has_var(str(var_name))
+        else:
+            return block_desc.has_var_recursive(str(var_name))
+
+    def _find_var(self, block_desc, var_name, is_forward):
+        if is_forward:
+            return block_desc.find_var(str(var_name))
+        else:
+            return block_desc.find_var_recursive(str(var_name))
+
     def memory_optimize(self):
         self._build_graph()
         self._dataflow_analyze()
         self.pool = []
         for i in range(self.op_size):
+            op = self._ops[i]
+            if op.type() == "while" or op.type() == "while_grad":
+                continue
+            block_desc = op.block()
+            is_forward = i < self._forward_num
             if self.pool:
-                out_pair = [(x, self.global_block_desc.var(str(x)).shape())
-                            for x in self._defs[i]]
+                defs_can_optimize = filter(
+                    lambda x: str(x) != "@EMPTY@" and self._has_var(block_desc, x, is_forward),
+                    self._defs[i])
+                defs_can_optimize = filter(
+                    lambda x: not self._find_var(block_desc, x, is_forward).persistable(),
+                    defs_can_optimize)
+                defs_can_optimize = filter(
+                    lambda x: self._find_var(block_desc, x, is_forward).type() == core.VarDesc.VarType.LOD_TENSOR,
+                    defs_can_optimize)
```

Comment:
Why not combine these three `filters` into one?

Reply:
Because I find that yapf can not format lambda style. It will be 

```python
defs_can_optimize = filter(
       lambda x: str(x) != "@EMPTY@" and self._has_var(block_desc, x, is_forward) and   not self._find_var(block_desc, x, is_forward).persistable() and self._find_var(block_desc, x, is_forward).type() == core.VarDesc.VarType.LOD_TENSOR
       self._defs[i])
```

It's too looooog!